File: backend-php/dao/UserDAO.php
<?php
/**
 * DAO : accès à la table users.
 * Équivalent PHP de com.ihecride.dao.UserDAO.
 */

require_once __DIR__ . '/../utils/DBConnection.php';
require_once __DIR__ . '/../models/User.php';

class UserDAO {

    public function findByEmail(?string $email): ?User {
        if ($email === null) return null;
        $stmt = DBConnection::getConnection()->prepare("SELECT * FROM users WHERE email = ?");
        $stmt->execute([$email]);
        $row = $stmt->fetch();
        return $row ? $this->map($row) : null;
    }

    public function findById(int $id): ?User {
        $stmt = DBConnection::getConnection()->prepare("SELECT * FROM users WHERE id = ?");
        $stmt->execute([$id]);
        $row = $stmt->fetch();
        return $row ? $this->map($row) : null;
    }

    public function insert(User $u): int {
        $sql = "INSERT INTO users(nom, prenom, email, telephone, mot_de_passe_hash, role) VALUES(?,?,?,?,?,?)";
        $pdo = DBConnection::getConnection();
        try {
            $stmt = $pdo->prepare($sql);
            $ok = $stmt->execute([
                $u->nom, $u->prenom, $u->email, $u->telephone,
                $u->motDePasseHash, $u->role
            ]);
            if ($ok && $stmt->rowCount() > 0) {
                return (int) $pdo->lastInsertId();
            }
        } catch (PDOException $e) {
            error_log('UserDAO::insert ' . $e->getMessage());
        }
        return -1;
    }

    /** @return User[] */
    public function findByRole(string $role): array {
        $stmt = DBConnection::getConnection()->prepare(
            "SELECT * FROM users WHERE role = ? ORDER BY created_at DESC"
        );
        $stmt->execute([$role]);
        $list = [];
        foreach ($stmt->fetchAll() as $row) $list[] = $this->map($row);
        return $list;
    }

    /** @return User[] */
    public function findAll(): array {
        $stmt = DBConnection::getConnection()->query("SELECT * FROM users ORDER BY created_at DESC");
        $list = [];
        foreach ($stmt->fetchAll() as $row) $list[] = $this->map($row);
        return $list;
    }

    public function updateStatut(int $id, string $statut): bool {
        $stmt = DBConnection::getConnection()->prepare("UPDATE users SET statut = ? WHERE id = ?");
        return $stmt->execute([$statut, $id]) && $stmt->rowCount() > 0;
    }

    public function update(User $u): bool {
        $stmt = DBConnection::getConnection()->prepare(
            "UPDATE users SET nom=?, prenom=?, telephone=?, photo=?, contact_confiance=? WHERE id=?"
        );
        return $stmt->execute([
            $u->nom, $u->prenom, $u->telephone, $u->photo, $u->contactConfiance, $u->id
        ]) && $stmt->rowCount() > 0;
    }

    public function delete(int $id): bool {
        $stmt = DBConnection::getConnection()->prepare("DELETE FROM users WHERE id = ?");
        return $stmt->execute([$id]) && $stmt->rowCount() > 0;
    }

    public function updateNoteMoyenne(int $userId): bool {
        $stmt = DBConnection::getConnection()->prepare(
            "UPDATE users SET note_moyenne = (SELECT AVG(note) FROM evaluations WHERE evalue_id = ?) WHERE id = ?"
        );
        return $stmt->execute([$userId, $userId]) && $stmt->rowCount() > 0;
    }

    private function map(array $rs): User {
        $u = new User();
        $u->id                = isset($rs['id']) ? (int) $rs['id'] : null;
        $u->nom               = $rs['nom'] ?? null;
        $u->prenom            = $rs['prenom'] ?? null;
        $u->email             = $rs['email'] ?? null;
        $u->telephone         = $rs['telephone'] ?? null;
        $u->motDePasseHash    = $rs['mot_de_passe_hash'] ?? null;
        $u->photo             = $rs['photo'] ?? null;
        $u->role              = $rs['role'] ?? null;
        $u->noteMoyenne       = isset($rs['note_moyenne']) ? (float) $rs['note_moyenne'] : 0.0;
        $u->statut            = $rs['statut'] ?? null;
        $u->contactConfiance  = $rs['contact_confiance'] ?? null;
        $u->createdAt         = $rs['created_at'] ?? null;
        return $u;
    }
}

File: backend/src/main/java/com/ihecride/dao/UserDAO.java
package com.ihecride.dao;

import com.ihecride.models.User;
import com.ihecride.utils.DBConnection;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;

/**
 * DAO : accès à la table users.
 */
public class UserDAO {

    public User findByEmail(String email) {
        String sql = "SELECT * FROM users WHERE email = ?";
        try (Connection c = DBConnection.getConnection();
             PreparedStatement ps = c.prepareStatement(sql)) {
            ps.setString(1, email);
            try (ResultSet rs = ps.executeQuery()) {
                if (rs.next()) return map(rs);
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return null;
    }

    public User findById(int id) {
        String sql = "SELECT * FROM users WHERE id = ?";
        try (Connection c = DBConnection.getConnection();
             PreparedStatement ps = c.prepareStatement(sql)) {
            ps.setInt(1, id);
            try (ResultSet rs = ps.executeQuery()) {
                if (rs.next()) return map(rs);
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return null;
    }

    public int insert(User u) {
        String sql = "INSERT INTO users(nom, prenom, email, telephone, mot_de_passe_hash, role) VALUES(?,?,?,?,?,?)";
        try (Connection c = DBConnection.getConnection();
             PreparedStatement ps = c.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS)) {
            ps.setString(1, u.getNom());
            ps.setString(2, u.getPrenom());
            ps.setString(3, u.getEmail());
            ps.setString(4, u.getTelephone());
            ps.setString(5, u.getMotDePasseHash());
            ps.setString(6, u.getRole());
            if (ps.executeUpdate() > 0) {
                try (ResultSet keys = ps.getGeneratedKeys()) {
                    if (keys.next()) return keys.getInt(1);
                }
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return -1;
    }

    public List<User> findByRole(String role) {
        List<User> list = new ArrayList<>();
        String sql = "SELECT * FROM users WHERE role = ? ORDER BY created_at DESC";
        try (Connection c = DBConnection.getConnection();
             PreparedStatement ps = c.prepareStatement(sql)) {
            ps.setString(1, role);
            try (ResultSet rs = ps.executeQuery()) {
                while (rs.next()) list.add(map(rs));
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return list;
    }

    public List<User> findAll() {
        List<User> list = new ArrayList<>();
        String sql = "SELECT * FROM users ORDER BY created_at DESC";
        try (Connection c = DBConnection.getConnection();
             Statement st = c.createStatement();
             ResultSet rs = st.executeQuery(sql)) {
            while (rs.next()) list.add(map(rs));
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return list;
    }

    public boolean updateStatut(int id, String statut) {
        String sql = "UPDATE users SET statut = ? WHERE id = ?";
        try (Connection c = DBConnection.getConnection();
             PreparedStatement ps = c.prepareStatement(sql)) {
            ps.setString(1, statut);
            ps.setInt(2, id);
            return ps.executeUpdate() > 0;
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return false;
    }

    public boolean update(User u) {
        String sql = "UPDATE users SET nom=?, prenom=?, telephone=?, photo=?, contact_confiance=? WHERE id=?";
        try (Connection c = DBConnection.getConnection();
             PreparedStatement ps = c.prepareStatement(sql)) {
            ps.setString(1, u.getNom());
            ps.setString(2, u.getPrenom());
            ps.setString(3, u.getTelephone());
            ps.setString(4, u.getPhoto());
            ps.setString(5, u.getContactConfiance());
            ps.setInt(6, u.getId());
            return ps.executeUpdate() > 0;
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return false;
    }

    public boolean delete(int id) {
        String sql = "DELETE FROM users WHERE id = ?";
        try (Connection c = DBConnection.getConnection();
             PreparedStatement ps = c.prepareStatement(sql)) {
            ps.setInt(1, id);
            return ps.executeUpdate() > 0;
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return false;
    }

    public boolean updateNoteMoyenne(int userId) {
        String sql = "UPDATE users SET note_moyenne = (SELECT AVG(note) FROM evaluations WHERE evalue_id = ?) WHERE id = ?";
        try (Connection c = DBConnection.getConnection();
             PreparedStatement ps = c.prepareStatement(sql)) {
            ps.setInt(1, userId);
            ps.setInt(2, userId);
            return ps.executeUpdate() > 0;
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return false;
    }

    private User map(ResultSet rs) throws SQLException {
        User u = new User();
        u.setId(rs.getInt("id"));
        u.setNom(rs.getString("nom"));
        u.setPrenom(rs.getString("prenom"));
        u.setEmail(rs.getString("email"));
        u.setTelephone(rs.getString("telephone"));
        u.setMotDePasseHash(rs.getString("mot_de_passe_hash"));
        u.setPhoto(rs.getString("photo"));
        u.setRole(rs.getString("role"));
        u.setNoteMoyenne(rs.getDouble("note_moyenne"));
        u.setStatut(rs.getString("statut"));
        u.setContactConfiance(rs.getString("contact_confiance"));
        u.setCreatedAt(rs.getTimestamp("created_at"));
        return u;
    }
}
